### DISCUSSION TITLE
[Error in running test_ros.py] ModuleNotFoundError: No module named 'PyKDL'

### DISCUSSION THREAD

**Discussion by ReanFernandes** [USC-ACTLab/crazyswarm#713] [Q&A]
While running all the instructions as directed on the physical robots and simulation installation page, I keep running into this error while running the $CSW_PYTHON -m pytest command. I have looked online for solutions to this problem, but insofar no methods have succeeded. The following is the error message i keep getting. Can someone please give me insight into how to solve this problem? I have tried both sudo apt install python3-PyKDL and pip install PyKDL commands but they both say that the requirement is already satifsfied. Thank you very much in advance :)

`platform linux -- Python 3.7.15, pytest-7.1.2, pluggy-1.0.0
rootdir: /home/rean/flight_control_lab/crazyswarm/ros_ws/src/crazyswarm/scripts, configfile: pytest.ini
collected 31 items                                                             

test_collisionAvoidance.py ............                                  [ 38%]
test_highLevel.py .........                                              [ 67%]
test_lowLevel.py ......                                                  [ 87%]
test_ros.py F                                                            [ 90%]
test_simOnly.py .                                                        [ 93%]
test_videoOutput.py s                                                    [ 96%]
test_yamlString.py .                                                     [100%]

=================================== FAILURES ===================================
_______________________________ test_ros_import ________________________________

    @pytest.mark.ros
    def test_ros_import():
>       from pycrazyswarm.crazyflie import TimeHelper, CrazyflieServer

test_ros.py:6: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pycrazyswarm/crazyflie.py:9: in <module>
    import tf_conversions
/opt/ros/noetic/lib/python3/dist-packages/tf_conversions/__init__.py:30: in <module>
    from .posemath import *
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    from geometry_msgs.msg import Pose, Point, Quaternion
    from tf import transformations
    import tf
    import rospy
    import numpy
    
>   from PyKDL import *
E   ModuleNotFoundError: No module named 'PyKDL'

/opt/ros/noetic/lib/python3/dist-packages/tf_conversions/posemath.py:34: ModuleNotFoundError
=============================== warnings summary ===============================
test_ros.py::test_ros_import
  /opt/ros/noetic/lib/python3/dist-packages/actionlib/goal_id_generator.py:49: DeprecationWarning: invalid escape sequence \p
    """

test_ros.py::test_ros_import
  /opt/ros/noetic/lib/python3/dist-packages/actionlib/goal_id_generator.py:59: DeprecationWarning: invalid escape sequence \p
    """

test_ros.py::test_ros_import
  /opt/ros/noetic/lib/python3/dist-packages/tf/transformations.py:167: DeprecationWarning: invalid escape sequence \*
    """

test_ros.py::test_ros_import
  /opt/ros/noetic/lib/python3/dist-packages/tf/transformations.py:912: DeprecationWarning: invalid escape sequence \*
    """

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED test_ros.py::test_ros_import - ModuleNotFoundError: No module named 'P...
============= 1 failed, 29 passed, 1 skipped, 4 warnings in 1.86s =============`

**Comment by knmcguire** [comment]
Are you sure that ROS is using the same python as the one you installed your module for? Do you have multiple Python distributions installed on your system?

**Reply by ReanFernandes** [reply]
Yes it turns out that i do.
My crazyswarm environment runs on 3.7.15.
The base environment runs on 3.9.12
and deactivating all environment gives me 2.7.18.
I will try uninstalling the others and make sure that i have only 3.7.15 and see if it works out. Thank you for the reply! :)

**Reply by mengou8023** [reply]
hello bro，have you dealed with this problem？I meet the same question now.
